### PR TITLE
JAVA-2371: Handle null elements in collections on the decode path

### PR DESCRIPTION
**Pull request by olim7t**

### CODE REVIEW

**Review by jtgrabowski — APPROVED** [review]

File: core/src/main/java/com/datastax/oss/driver/internal/core/type/codec/MapCodec.java [hunk 1]
```diff
@@ -134,18 +134,28 @@ public ByteBuffer encode(
       int size = input.getInt();
       Map<KeyT, ValueT> result = Maps.newLinkedHashMapWithExpectedSize(size);
       for (int i = 0; i < size; i++) {
+        KeyT key;
```
Comment (tomekl007) [inline core/src/main/java/com/datastax/oss/driver/internal/core/type/codec/MapCodec.java:137]: If we are handling Map shouldn't we handle that for Tuple and Udt as well?

File: core/src/test/java/com/datastax/oss/driver/internal/core/type/codec/MapCodecTest.java [hunk 2]
```diff
@@ -118,6 +118,16 @@ public void should_decode_non_empty_map() {
...
   }
 
+  @Test
+  public void should_decode_map_with_null_elements() {
+    when(keyCodec.decode(Bytes.fromHexString("0x10"), ProtocolVersion.DEFAULT)).thenReturn("a");
+    when(valueCodec.decode(Bytes.fromHexString("0x0002"), ProtocolVersion.DEFAULT)).thenReturn(2);
+    assertThat(decode("0x" + "00000002" + "0000000110" + "FFFFFFFF" + "FFFFFFFF" + "000000020002"))
+        .containsOnlyKeys("a", null)
+        .containsEntry("a", null)
+        .containsEntry(null, 2);
```
Comment (tomekl007) [inline core/src/test/java/com/datastax/oss/driver/internal/core/type/codec/MapCodecTest.java:128]: I think that we are missing a test case for key and value null for the same entry but I don't know if such a test gives much value.

Comment (olim7t) [reply core/src/test/java/com/datastax/oss/driver/internal/core/type/codec/MapCodecTest.java:128]: It wouldn't increase coverage so I would say no.

File: core/src/main/java/com/datastax/oss/driver/internal/core/type/codec/ListCodec.java [hunk 3]
```diff
@@ -115,11 +115,19 @@ public ByteBuffer encode(
...
       List<ElementT> result = new ArrayList<>(size);
       for (int i = 0; i < size; i++) {
+        ElementT element;
         int elementSize = input.getInt();
-        ByteBuffer encodedElement = input.slice();
-        encodedElement.limit(elementSize);
-        input.position(input.position() + elementSize);
-        result.add(elementCodec.decode(encodedElement, protocolVersion));
+        // Allow null elements on the decode path, because Cassandra might return such collections
+        // for some computed values in the future -- e.g. SELECT ttl(some_collection)
```
Comment (adutra) [inline core/src/main/java/com/datastax/oss/driver/internal/core/type/codec/ListCodec.java:121]: I'm probably a bit paranoid but should we include the example of `ttl(collection)`? This example is DSE-specific.

File: core/src/main/java/com/datastax/oss/driver/internal/core/type/codec/ListCodec.java [hunk 4]
```diff
@@ -115,11 +115,19 @@ public ByteBuffer encode(
...
+        if (elementSize < 0) {
+          element = null;
+        } else {
+          ByteBuffer encodedElement = input.slice();
+          encodedElement.limit(elementSize);
+          input.position(input.position() + elementSize);
```
Comment (adutra) [inline core/src/main/java/com/datastax/oss/driver/internal/core/type/codec/ListCodec.java]: Nit: it would be a bit easier to understand this code if you invert these two lines:

```
element = elementCodec.decode(encodedElement, protocolVersion);
input.position(input.position() + elementSize);
```

(I initially thought that updating the position of `input` was required before the element can be decoded.)

File: core/src/test/java/com/datastax/oss/driver/internal/core/type/codec/ListCodecTest.java [hunk 5]
```diff
@@ -93,6 +93,12 @@ public void should_decode_non_empty_list() {
         .containsExactly(1, 2, 3);
   }
 
+  @Test
+  public void should_decode_list_with_null_elements() {
+    when(elementCodec.decode(Bytes.fromHexString("0x0002"), ProtocolVersion.DEFAULT)).thenReturn(2);
+    assertThat(decode("0x" + "00000002" + "FFFFFFFF" + "000000020002")).containsExactly(null, 2);
```
Comment (adutra) [inline core/src/test/java/com/datastax/oss/driver/internal/core/type/codec/ListCodecTest.java]: This is very cryptic, how about:

```
assertThat(
        decode(
            "0x"
                + "00000002" // number of elements
                + "FFFFFFFF" // size of element 1 (null)
                + "00000002" // size of element 2 (2 bytes)
                + "0002" // element 2
            ))
    .containsExactly(null, 2);
```

**Review by adutra — CHANGES_REQUESTED** [review]

File: core/src/main/java/com/datastax/oss/driver/internal/core/type/codec/MapCodec.java (hunk 1 above)
Comment (olim7t) [reply core/src/main/java/com/datastax/oss/driver/internal/core/type/codec/MapCodec.java:137]: `TupleCodec` and `UdtCodec` already handle null values.

UDT field names don't appear in the encoded representation. And there's no way they can ever be null anyway.

File: core/src/main/java/com/datastax/oss/driver/internal/core/type/codec/ListCodec.java (hunk 3 above)
Comment (olim7t) [reply core/src/main/java/com/datastax/oss/driver/internal/core/type/codec/ListCodec.java:121]: [CASSANDRA-8877](https://issues.apache.org/jira/browse/CASSANDRA-8877) will likely add that feature.

File: core/src/main/java/com/datastax/oss/driver/internal/core/type/codec/ListCodec.java (hunk 4 above)
Comment (olim7t) [reply core/src/main/java/com/datastax/oss/driver/internal/core/type/codec/ListCodec.java]: ok I'm doing it in a separate commit

File: core/src/test/java/com/datastax/oss/driver/internal/core/type/codec/ListCodecTest.java (hunk 5 above)
Comment (olim7t) [reply core/src/test/java/com/datastax/oss/driver/internal/core/type/codec/ListCodecTest.java]: There's a commented example 30 lines above. When you have the whole file open in an IDE it's pretty visible, I think it's fine not to repeat it in every test.

Comment (tomekl007) [reply core/src/test/java/com/datastax/oss/driver/internal/core/type/codec/ListCodecTest.java]: I agree with Alex on that, I had the same impression that it was very cryptic

Comment (adutra) [reply core/src/test/java/com/datastax/oss/driver/internal/core/type/codec/ListCodecTest.java]: Well it's not exactly the same example, here we have a `"FFFFFFFF"` block that seems strange at first glance, and `"000000020002"` is also strange since it contains both the size of element 2 and element 2 itself.
